### ISSUE TITLE
Fix All Compiler Warnings

### ISSUE THREAD

**Issue by garethellis0** [UBC-Thunderbots/Software#779]
### Description of the task
With our migration to bazel, we have started to get a _lot_ of compiler warnings that we just didn't have turned on before. We should fix them all and make warnings fail compilation.

### Acceptance criteria

- [ ] Make warnings errors (so they will fail compilation)
- [ ] Fix all warnings in the codebase

**Comment by garethellis0** [comment]
While updating some toolchain stuff for https://github.com/UBC-Thunderbots/Software/issues/849, namely updating our compiler to clang instead of gcc. I'm getting some warnings that should definitely have caused errors with `-Werror`. I _think_ what might have happened is that some cached build outputs were not properly rebuilt when the flags were changed. @scveloso, can you:
- disable the remote cache by commenting out the line `build --remote_http_cache=https://storage.googleapis.com/thunderbots_bazel_cache_us_multi_region` in `src/.bazelrc`
- run `bazel clean --expunge` in `src`
and then see if there are any compiler errors that we should fix? If so we should re-open this.

**Comment by garethellis0** [comment]
One example is that `point` is a unused variable on line 171 of `polygon_test.cpp`:
```
TEST(PolygonTest, test_polygon_triangle_intersects_ray2)
{
    Point p1{0.0f, 0.0f}, p2{1.0f, 0.0f}, p3{1.0f, 0.0f};
    Point point{2.0f, 2.0f};
    Polygon triangle{p1, p2, p3};

    Ray ray{Point(-0.1f, -0.1f), Vector(-1.0f, -1.0f)};
    EXPECT_FALSE(triangle.intersects(ray));
}
```

**Comment by garethellis0** [comment]
Ok, in order to limit scope on #849 (it's already really big), current scope for this ticket is as follows:
- disable all caching and clear the bazel cache, as detailed above
- enable warnings as errors by adding `"-Werror"` to the `warnings_feature` in `cc_toolchain_config.bzl`
- enable the `-Winvalid-constexpr` warning by removing `-Wno-invalid-constexpr` from the `warnings_feature` in `cc_toolchain_config.bzl`
- figure out how to disable warnings for external packages, like `gtest`, `boost`, `g3log`, etc.

This ticket will have to be blocked on #849 as well.

@scveloso - if this seems reasonable to you, could you please update the description for this issue accordingly and re-open this issue?

**Comment by garethellis0** [comment]
It's also possible we've gained a few more error flags in the `gcc`->`clang` migration. If that's the case then we can probably file a new ticket, since this one was just for the `gcc` errors.

**Comment by scveloso** [comment]
@garethellis0 this is reasonable. I'll take a closer look later and update the issue as soon as possible.

**Comment by garethellis0** [comment]
Ok, looks like the new toolchain catches some issues the old one doesn't. My comment above stands, re-opening.

@scveloso - no need for you to work on this right now, would rather you focus on geom, this can go back into the ticket pool.

**Comment by garethellis0** [comment]
@amrm3lm - just as an additional note, just do a quick search over the codebase for this issue number (779), I think we've tagged a few places stuff needs to be changed once this is fixed.
